sync/pooled_worker_pool.go

```diff
@@ -0,0 +1,122 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package sync implements synchronization facililites such as worker pools.
+package sync
+
+import "math/rand"
+
+const (
+	defaultNumShards = 10
+)
+
+var (
+	// Var for testing purposes
+	defaultKillWorkerProbability = 0.0001
+)
+
+// PooledWorkerPool provides a pool for goroutines, but unlike WorkerPool,
+// the actual goroutines themselves are re-used. This can be useful from a
+// performance perspective in scenarios where the allocation and growth of
+// the new goroutine and its stack is a bottleneck. Specifically, if the
+// work function being performed has a very deep call-stack, calls to
+// runtime.morestack can dominate the workload. Re-using existing goroutines
+// allows the stack to be grown once, and then re-used for many invocations.
+//
+// In order to prevent abnormally large goroutine stacks from persisting over
+// the life-cycle of an application, the PooledWorkerPool will randomly kill
+// existing goroutines and spawn a new one.
+//
+// The PooledWorkerPool also implements sharding of its underlying worker channels
+// to prevent excessive lock contention.
+type PooledWorkerPool interface {
+	// Init initializes the pool.
+	Init()
+
+	// Go waits until the next worker becomes available and executes it.
+	Go(work Work)
+}
+
+type pooledWorkerPool struct {
+	workChs   []chan workerInstruction
+	numShards int
+}
+
+type workerInstruction struct {
+	work      Work
+	shouldDie bool
+}
+
+// NewPooledWorkerPool creates a new worker pool.
+func NewPooledWorkerPool(size int) PooledWorkerPool {
+	numShards := defaultNumShards
+	if size < numShards {
+		numShards = size
+	}
+
+	workChs := make([]chan workerInstruction, numShards)
+	for i := range workChs {
+		workChs[i] = make(chan workerInstruction, size/numShards)
+	}
+
+	return &pooledWorkerPool{
+		workChs:   workChs,
+		numShards: numShards,
+	}
+}
+
+func (p *pooledWorkerPool) Init() {
+	for _, workCh := range p.workChs {
+		for i := 0; i < cap(workCh); i++ {
+			p.spawnWorker(workCh)
+		}
+	}
+}
+
+func (p *pooledWorkerPool) Go(work Work) {
+	instruction := workerInstruction{work: work}
+	randVal := rand.Float64()
```

Comment:
Do you want to create a new rand and keep inside the pooledWorkerPool struct instead of using the globalRand?

Reply:
done